client/app/services/query.js

```diff
@@ -454,26 +454,30 @@ function QueryResource(
       });
     }
 
-    if (parameters.isRequired()) {
-      // Need to clear latest results, to make sure we don't use results for different params.
-      this.latest_query_data = null;
-      this.latest_query_data_id = null;
-    }
-
-    if (this.latest_query_data && maxAge !== 0) {
-      if (!this.queryResult) {
-        this.queryResult = new QueryResult({
-          query_result: this.latest_query_data,
-        });
+    if (isDirty) {
+      if (parameters.isRequired()) {
+        // Need to clear latest results, to make sure we don't use results for different params.
+        this.latest_query_data = null;
+        this.latest_query_data_id = null;
       }
-    } else if (this.latest_query_data_id && maxAge !== 0) {
-      if (!this.queryResult) {
-        this.queryResult = QueryResult.getById(this.latest_query_data_id);
+
+      if (this.latest_query_data && maxAge !== 0) {
+        if (!this.queryResult) {
+          this.queryResult = new QueryResult({
+            query_result: this.latest_query_data,
+          });
+        }
+      } else if (this.latest_query_data_id && maxAge !== 0) {
+        if (!this.queryResult) {
+          this.queryResult = QueryResult.getById(this.latest_query_data_id);
+        }
+      } else if (this.data_source_id) {
+        this.queryResult = QueryResult.get(this.data_source_id, queryText, parameters.getValues(), maxAge, this.id);
+      } else {
+        return new QueryResultError('Please select data source to run this query.');
       }
-    } else if (this.data_source_id) {
-      this.queryResult = QueryResult.get(this.data_source_id, queryText, parameters.getValues(), maxAge, this.id);
     } else {
-      return new QueryResultError('Please select data source to run this query.');
+      this.queryResult = QueryResult.getByQueryId(this.id, parameters.getValues(), maxAge);
```

Comment:
Not sure actually, because of the parameters. :-\ But we need to fully review how this works, because it has implications in dashboards. Maybe in other places, but iirc this was added for the benefit of widgets.

This is one of the oldest pieces of code in the UI codebase that survived many rewrites, but overtime mutated into a monster.